### ISSUE TITLE
Use condition's variable in the error_message of a validation block

### ISSUE THREAD

**Issue by C-Duv** [hashicorp/terraform#35775]
### Terraform Version

```shell
Terraform v1.9.6
on linux_amd64
```

### Use Cases

*tl;dr: I want to use the current value of a runtime variable created in the `condition` part of a variable `validation`, to give more details about the error.*

I am looking for a way to improve my `error_message` for variable validation in my Terraform project.

Say I have 2 variable:

* `availabilityzones` gives the list of AZ my module/project can use.
* `workloads` defines some workload to start for this module/project can use.

Workloads have an AMI (not relevant here) and are given a number of AZ they can use (span): `az_count`.

My `workloads` variable has a validation to make sure the number of AZ is consistent with the size of `availabilityzones`:

```terraform
condition = alltrue([
  for workload_name, workload in var.workloads :
    1 <= workload.az_count && workload.az_count <= length(var.availabilityzones)
])
```

Validation works and can detect too large `az_count`, but I would like to tell the user which workload caused the issue, so I tried to use the `workload_name` variable (the one from the `condition`) in the `error_message`:

```terraform
error_message = "For a workload, the AZ count must be between `1` and the number of configured AZ (both inclusive). Workload \"${workload_name}\" has: ${tonumber(workload.az_count)}"
```

But it fails:

```
│ Error: Invalid reference
│ 
│   on …/variables.tf line 42, in variable "workloads":
│  129:     error_message = "For a workload, the AZ count must be between `1` and the number of configured AZ (both inclusive). Workload \"${workload_name}\" has: ${tonumber(workload.az_count)}"
│ 
│ A reference to a resource type must be followed by at least one attribute access, specifying the resource name.
```

I am guessing the context from `condition` is lost once validation has detected an error, and it not passed to `error_message`'s context.

<details>
    <summary>Full code</summary>

```terraform
variable "availabilityzones" {
  description = "The availability zones (AZs) to use."
  default = [
    "eu-west-1a",
    "eu-west-1b",
    "eu-west-1c",
  ]
  type = set(string)
}

variable "workloads" {
  default = {
    workload_foo = {
      az_count = 1
      ami = "ami-061ff9349538cd3f1"
    }
    workload_bar = {
      az_count = 2
      ami = "ami-0e309138dd3e25b2e"
    }
    workload_bad = {
      az_count = 4 # ⟽ Too many AZ here
      ami = "ami-040a86d5d7bf29267"
    }
  }
  type = map(object({
      # How many AZ to use?
      az_count = number

      # Which AMI should the workload use?
      ami = string
    }))

  validation {
    condition = alltrue([
      for workload_name, workload in var.workloads :
        1 <= workload.az_count && workload.az_count <= length(var.availabilityzones)
    ])
    error_message = "For a workload, the AZ count must be between `1` and the number of configured AZ (both inclusive). Workload \"${workload_name}\" has: ${tonumber(workload.az_count)}"
  }
}
```
</details>

### Attempted Solutions

A workaround I could think of would be repeating the test in `error_message` to re-find the invalid value, but this would always find/display the first invalid value, even if multiple workloads are incorrect.

Another was to store each check in a map and assert if it's empty, otherwise display it's content. This was suggested on StackOverflow by [Paolo](https://stackoverflow.com/users/3390419/paolo):

```terraform
variable "availabilityzones" {
  description = "The availability zones (AZs) to use."
  default = [
    "eu-west-1a",
    "eu-west-1b",
    "eu-west-1c",
  ]
  type = set(string)
}

variable "workloads" {
  default = {
    workload_foo = {
      az_count = 1
      ami      = "ami-061ff9349538cd3f1"
    }
    workload_bar = {
      az_count = 2
      ami      = "ami-0e309138dd3e25b2e"
    }
    workload_bad = {
      az_count = 4 # ⟽ Too many AZ here
      ami      = "ami-040a86d5d7bf29267"
    }
  }
  type = map(object({
    # How many AZ to use?
    az_count = number

    # Which AMI should the workload use?
    ami = string
  }))

}

locals {
  workflow_check = {
    for workload_name, workload in var.workloads :
    workload_name => [1 <= workload.az_count && workload.az_count <= length(var.availabilityzones), workload.az_count]
  }
  invalid_workflow = {for k,v in local.workflow_check : k => tostring(v[1]) if v[0] == false}
}

check "validate_variables" {
  assert {
    condition = local.invalid_workflow == {}
    error_message = local.invalid_workflow == {} ? "" : "For a workload, the AZ count must be between `1` and the number of configured AZ (both inclusive). Workload \"${keys(local.invalid_workflow)[0]}\" has: ${tonumber(values(local.invalid_workflow)[0])}"
  }
}
```

Both solution seems _hacky_ and less readable and simply using the variable we have in the `validation` block.

### References

- StackOverflow question prior to this issue: [_Can I use condition's variable in the error_message of a Terraform variable validation?_](https://stackoverflow.com/questions/78981520/can-i-use-conditions-variable-in-the-error-message-of-a-terraform-variable-vali/)

**Comment by jbardin** [comment]
Hi @C-Duv,

What you're asking for isn't possible for a couple reasons. As you've guessed, the `workload` is scoped to within the for expression and does not exist outside of that context. Being able to reference a nested symbol like that would require ensuring all symbols used in the configuration are globally unique, which would be quite surprising for most users and cumbersome to workaround. Additionally, even if you would reference `workload`, the value is not going to be what you expect, if the value still did exist it would technically be the final value used within the for expression, completely unrelated to the result of of the `alltrue` function. The `alltrue` function is receiving a list of boolean values and has no idea what expression generated those values, just as `condition` is receiving a single boolean value and has no idea what expression generated that value.

If you need to extract a condition from a collection of values, then processing that value in multiple steps is going to be the best solution for now. Note however that your `invalid_workflow == {}` is not correct, because the type of `invalid_workflow` can never be equal to the type of "empty object". Use `length` to check for an empty collection.

It would probably be best to follow #34428 for any future enhancements around validation complex data structures.
